### DISCUSSION TITLE
Side by side images

### DISCUSSION THREAD

**Discussion by wabiloo** [GuardKenzie/chafa.py#53] [General]
Hi,

I'm trying to find a reliable way to display 3 images side by side (horizontal grid) - they are all of the same size / aspect-ratio.
When the pixel mode is SYMBOLS (which seems to be the default that's decided by detect_capabilities() on my mac, both in VS Code (integrated terminal) and iTerm2 (not quite sure why it doesn't detect support for iTerm2 protocol), that's easy enough as I can extract the lines from the canvas, concatenate them, and print them line by line.

However when I force ITERM2 pixel format, that obviously cannot work.
I resulted to write some really clunky code that uses ANSI codes to move the cursor back up after printing each image, which kinda works. 

However, I'd like to find a way to automatically determine:
- What pixel mode was used
- On that basis, determine what mechanism to use to place the images side by side
- And in particular, determine how many up-cursor characters to print to get to the top of the previous image.

Any suggestion?

**Comment by GuardKenzie** [comment]
Hiya! This is quite a late reply, sorry for that.

I think the best way to achieve this is using ANSI codes, as you stated. You can now use the new [`TermInfo.emit`](https://chafapy.mage.black/api/TermInfo.html#chafa.TermInfo.emit) function to construct those.

You can check which pixel mode was used by just comparing [`CanvasConfig.pixel_mode`](https://chafapy.mage.black/api/CanvasConfig.html#chafa.ReadOnlyCanvasConfig.pixel_mode) to the enumerator values in [`PixelMode`](https://chafapy.mage.black/api/enums.html#PixelMode).

Since when you specify the canvas config height and width, you know the size of the area your image is going to fit into.

The only tricky part is that the amounts you need to move to the right and upwards after printing an image with each graphics protocol varies. For sixels, you need to move up and to the right `CanvasConfig.height` and `CanvasConfig.width` respectively, however, when using ITERM2 (and kitty), you need to move up `CanvasConfig.height - 1` lines and you don't need to move right.

There is currently a bug in 1.2.0 where `TermInfo.detect_capabilities` doesn't detect ITERM2 image protocol support (my bad, I forgot to check for it in the code). I have fixed this in eefcb1f8cf02b2aa0c6414c540c072963c68b2fb. You can remedy this by checking `TermInfo.have_seq(TermSeq.CHAFA_TERM_SEQ_BEGIN_ITERM2_IMAGE)` manually.
